### PR TITLE
Fix issues around type directed conversion

### LINKED ISSUE #12994: Error for extra implicit conversion is needed
Right now implicit type conversion happens only one-step (like `int-> float` or `float -> Nullable<float>`), but when we try to do two-step conversion (e.g. `int -> Nullable<float>`) there is no warning on compile time and we've exception in runtime.

**Repro steps**

```fsharp
type C() =
    member val X = System.Nullable(1.0) with get, set
C(X = 1)
```

**Expected behavior**

The code either works or gives compiler warning

**Actual behavior**

The code compiles successfully but in runtime the exception is raised 

```
Unhandled exception. System.InvalidCastException: Unable to cast object of type 'System.Int32' to type 'System.Nullable`1[System.Double]'.
```

**Known workarounds**

Fix the code (e.g. write C(X = 1.0)) after facing the error in the runtime

**Related information**

dotnet 6.0.201
fsharp 6.0.3

### PR DESCRIPTION
**Pull request by NinoFloris**
This PR is split into a few parts:

1. Tests, some fail because of #12994 while another test fails expecting a ctor call while getting op_Implicit due to the code falling back to [FS-1093](https://github.com/fsharp/fslang-design/blob/main/FSharp-6.0/FS-1093-additional-conversions.md).
2. Fix for #12994, copying what @dsyme did in #12423.
3. Improved diagnostics by keeping the known parameter type around instead of reporting the underlying type. This results in a message going from `The type 'bool' is not compatible with type 'float'` to `The type 'bool' is not compatible with type 'Nullable<float>'` which is what users would expect to see.
4. A bit more contentious, it integrates the fix from 2 into the type directed conversions code of FS-1093, completing a nullable TODO comment and removing the old superseded code.

Why integrate with the code for [FS-1093](https://github.com/fsharp/fslang-design/blob/main/FSharp-6.0/FS-1093-additional-conversions.md)? De-facto it already is; op_Implicit would trigger if [FS-1075](https://github.com/fsharp/fslang-design/blob/main/FSharp-5.0/FS-1075-nullable-interop.md) would not exist in most cases (except the double conversion ones).
- This is most clear when looking at #12994 which was missed in the *implementation* of FS-1075; Passing a setter argument `float` to `Nullable<float>` results in `op_Implicit` without a warning being emitted today.
- Additionally having FS-1075 overlap FS-1093 means it obscures the diagnostics introduced in the latter due to it not being seen as a type directed conversion (which means it does not produce consistent warnings for System.Nullable, see https://github.com/dotnet/fsharp/issues/13437). It now produces a builtin diag when it's a method arg (something to discuss) and an implicit conversion diag for the others.

@dsyme Maybe this needs a new RFC or change to FS-1093? Adding the diag seemed so small to me that it would fit under tidying up things.

In neither implementation nothing beyond the diag has been widened past what was specified in FS-1093 and FS-1075, notably:
- Double conversions for System.Nullable are still restricted to method arguments only.
- Type directed conversions for System.Nullable outside method args still fall through to op_Implicit, meaning that we still error on `let x: Nullable<float> = 1: int` as expected.
- Setter arguments produce a constructor call instead of an op_Implicit as an expected change to fix #12994.

@dsyme related, looking at line https://github.com/dotnet/fsharp/blob/9f65a3b6d3c6f6171e297fe55dbec1c42b235009/src/Compiler/Checking/MethodCalls.fs#L1306 if the compiler runs into a conversion that it cannot properly do it just spews `box unbox.any` due to the coerce. I assume this only happened because the type checks were more lenient than the expr adjustments?
Also, none of the nullable interop conversions in `AdjustCallerArgForOptional` were guarded by the language feature flag, it is now guarded in `AdjustExprForTypeDirectedConversions`.  Maybe it was ok because it was already checked under `AdjustCalledArgTypeForOptionals`?

### CODE REVIEW

**Comment by NinoFloris** [comment]
Maybe this should be combined with similar work in #13063?

**Comment by NinoFloris** [comment]
Fun test failure, the dumped IL is different on my machine by one dot, probably some ildasm culture output difference.

CI error:
```
 FSharp.Compiler.UnitTests.TypeDirectedConversionTests.float converts to System.Nullable<float> in method call parameter

==
Name: '.method public static void  test() cil managed'

Expected:	 IL_0000:  ldc.r8     100
Actual:		 IL_0000:  ldc.r8     100.
==
```

Changing it locally to include the `.` produces the inverse error. Moved these tests to use an int.

**Comment by dsyme** [comment]
@vzarytovskii @NinoFloris I'll need to review this very carefully.  I'll be back at work properly next week and suggest we do it then

File: src/Compiler/Checking/MethodCalls.fs [hunk 1]
```diff
@@ -279,16 +279,32 @@ let rec AdjustRequiredTypeForTypeDirectedConversions (infoReader: InfoReader) ad
...
+    elif g.langVersion.SupportsFeature LanguageFeature.NullableOptionalInterop && isMethodArg && isNullableTy g reqdTy && not (isNullableTy g actualTy) then 
+        let reqdTy2 =
+            if isTyparTy g reqdTy then
+                let tp = destTyparTy g reqdTy 
+                match tp.Constraints |> List.choose (function TyparConstraint.CoercesTo (c, _) -> Some c | _ -> None) with
+                | [reqdTy2] when tp.Rigidity = TyparRigidity.Flexible -> reqdTy2
+                | _ -> reqdTy
+            else reqdTy
+       
+        if not (isTyparTy g reqdTy2) &&
```
Comment (NinoFloris) [inline src/Compiler/Checking/MethodCalls.fs:301]: This could be simplified as the checks on the match arm already rule out reqdTy is a typar

File: src/Compiler/Checking/MethodCalls.fs [hunk 2]
```diff
@@ -279,16 +279,32 @@ let rec AdjustRequiredTypeForTypeDirectedConversions (infoReader: InfoReader) ad
...
+            (not (TypeFeasiblySubsumesType 0 g infoReader.amap m reqdTy2 CanCoerce actualTy) ||
+            isTyparTy g actualTy && (let ftyvs = freeInType CollectAll reqdTy2 in ftyvs.FreeTypars.Contains(destTyparTy g actualTy))) then
+
+            actualTy, TypeDirectedConversionUsed.Yes(warn TypeDirectedConversion.BuiltIn), Some (reqdTy, reqdTy2, ignore)
```
Comment (NinoFloris) [inline src/Compiler/Checking/MethodCalls.fs:305]: Regardless of missed simplification, this bit of code worries me. It's written in such a way it mirrors logic in TryFindRelevantImplicitConversion but originally NullableInterop did not ever return any new equational constraints. Maybe eqn here should always be None?

Comment (NinoFloris) [reply src/Compiler/Checking/MethodCalls.fs]: Ok I've removed it.

This was a holdover from earlier code when this arm was not protected by isMethodArg, where it would have needed to support the same solving complexity as the op_Implicit arm.

**Comment by NinoFloris** [comment]
I included the fix for #12946 as well (could easily be split out if preferred). I've written about the source of the current broken behavior here https://github.com/dotnet/fsharp/issues/12946#issuecomment-1217121581.

File: tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs [hunk 3]
```diff
@@ -0,0 +1,288 @@
+// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
```
Comment (vzarytovskii) [inline tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs:1]: More of a nitpick, but for new tests in future we probably prefer to have them in the ComponentTests suite, it's a bit easier to chain (i.e. baseline validation, IL validation, running, etc), and a bit easier to debug and configure. Not a big deal though.

File: tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+
+type Thing() =
+    static member Do(i: System.Nullable<float>) = ()
+    
+let test() = Thing.Do(true)
+            """
+            FSharpDiagnosticSeverity.Error
+            193
+            (7, 22, 7, 28)
+            """Type constraint mismatch. The type
```
Comment (NinoFloris) [inline tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs:196]: @dsyme this is a bit of an odd error IMO. 

Last I tested it has to do with the fact we create a type variable (this branch https://github.com/dotnet/fsharp/blob/ffa2e0ddb3f6ed8df0bf91a2e5c2c6a070b8c0f3/src/Compiler/Checking/MethodCalls.fs#L382) that then outputs this on a mismatch instead of `This expression was expected to have type ...`

It's not a big deal but still not the error you'd expect calling a fully concrete method.

**Comment by dsyme** [comment]
> if the compiler runs into a conversion that it cannot properly do it just spews box unbox.any due to the coerce. I assume this only happened because the type checks were more lenient than the expr adjustments

I think this case should match coercion type-drected conversions - which are originate in CheckExpressions.fs via MustCoerce and so on.

> Also, none of the nullable interop conversions in AdjustCallerArgForOptional were guarded by the language feature flag, it is now guarded in AdjustExprForTypeDirectedConversions. Maybe it was ok because it was already checked under AdjustCalledArgTypeForOptionals?

It might be (I need to double check) that these specific coercions were present in F# 5.0 for "nullable interop" i.e. https://github.com/fsharp/fslang-design/blob/main/FSharp-5.0/FS-1075-nullable-interop.md.  Or else yes, the checking in AdjustCalledArgTypeForOptionals should be sufficient - there is no specific need to guard again.

File: src/Compiler/Checking/MethodCalls.fs [hunk 5]
```diff
@@ -279,21 +282,33 @@ let rec AdjustRequiredTypeForTypeDirectedConversions (infoReader: InfoReader) ad
...
     elif g.langVersion.SupportsFeature LanguageFeature.AdditionalTypeDirectedConversions && typeEquiv g g.nativeint_ty reqdTy && typeEquiv g g.int32_ty actualTy then 
-       g.int32_ty, TypeDirectedConversionUsed.Yes(warn TypeDirectedConversion.BuiltIn), None
+        g.int32_ty, TypeDirectedConversionUsed.Yes(warn TypeDirectedConversion.BuiltIn, false), None
 
     // Adhoc int32 --> float64
     elif g.langVersion.SupportsFeature LanguageFeature.AdditionalTypeDirectedConversions && typeEquiv g g.float_ty reqdTy && typeEquiv g g.int32_ty actualTy then 
-       g.int32_ty, TypeDirectedConversionUsed.Yes(warn TypeDirectedConversion.BuiltIn), None
+        g.int32_ty, TypeDirectedConversionUsed.Yes(warn TypeDirectedConversion.BuiltIn, false), None
 
+    elif g.langVersion.SupportsFeature LanguageFeature.NullableOptionalInterop && isMethodArg && isNullableTy g reqdTy && not (isNullableTy g actualTy) then
```
Comment (dsyme) [inline src/Compiler/Checking/MethodCalls.fs:295]: Could you explain this additional rule a little more for me.   I see it's introducing the two-step conversion. 

Also why only `isMethodArg` here?  In the other rules, `isMethodArg` is only used to determine whether we give a warning.

File: src/Compiler/Checking/MethodCalls.fs [hunk 6]
```diff
@@ -352,9 +367,8 @@ let AdjustCalledArgTypeForOptionals (infoReader: InfoReader) ad enforceNullableO
 
                 // If inference has worked out it's a struct (e.g. an int) then use this
                 elif isStructTy g callerArgTy then
-                    let calledArgTy2 = destNullableTy g calledArgTy
-                    AdjustRequiredTypeForTypeDirectedConversions infoReader ad true false calledArgTy2 callerArgTy m
-
+                    AdjustRequiredTypeForTypeDirectedConversions infoReader ad true false calledArgTy callerArgTy m
```
Comment (dsyme) [inline src/Compiler/Checking/MethodCalls.fs:370]: I see, so `AdjustRequiredTypeForTypeDirectedConversions` is now doing the work to do the Nullable conversion

File: src/Compiler/Checking/MethodCalls.fs [hunk 7]
```diff
@@ -1521,21 +1527,8 @@ let AdjustCallerArgForOptional tcVal tcFieldInit eCallerMemberName (infoReader:
                         // This should be unreachable but the error will be reported elsewhere
                         callerArgExpr
                 else
-                    if isNullableTy g calledArgTy  then
```
Comment (dsyme) [inline src/Compiler/Checking/MethodCalls.fs:1524]: Quite a lot of logic removed here.  Does `AdjustCallerArgExpr` really cover all this?  Is it just a simplification or a fix?

File: src/Compiler/Checking/MethodCalls.fs [hunk 8]
```diff
@@ -1492,14 +1503,9 @@ let AdjustCallerArgForOptional tcVal tcFieldInit eCallerMemberName (infoReader:
             | NotOptional ->
                 //  T --> Nullable<T> widening at callsites
                 if isOptCallerArg then errorR(Error(FSComp.SR.tcFormalArgumentIsNotOptional(), m))
-                if isNullableTy g calledArgTy then
```
Comment (dsyme) [inline src/Compiler/Checking/MethodCalls.fs:1495]: Again, quite a lot of logic removed here. Does AdjustCallerArgExpr really cover all this? Is it just a simplification or a fix?

File: src/Compiler/Checking/MethodCalls.fs (hunk 5 above)
Comment (NinoFloris) [reply src/Compiler/Checking/MethodCalls.fs:295]: This rule is mostly transplanting checks that existed on the relevant call sites to make sure it only hits in the same cases as the callers used to. Important however for the behavior from both RFCs is to keep this conversion limited to method args only.

I wrote about that in the PR
>Type directed conversions for System.Nullable outside method args still fall through to op_Implicit, meaning that we still error on let x: Nullable<float> = 1: int as expected.

We don't want to enable double conversions for these cases as it was never specced to be allowed.

Beyond that though, the two step conversion was always there and called out in FS-1093 (starting from here https://github.com/fsharp/fslang-design/blob/main/FSharp-6.0/FS-1093-additional-conversions.md#interaction-with-optional-arguments). 

It was just split between ad-hoc checks outside that called into the TDC with the nullable underlying type. (see https://github.com/dotnet/fsharp/pull/13673/files#diff-33f19e1cff6e1df09e36df423722449340f685c5c08f041543334c173cc259f8L355)

File: tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs (hunk 4 above)
Comment (dsyme) [reply tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs:196]: Yes, I understand.  It's ok - AFAIK we can't remove the creation of that type variable, at least without affecting compat.  This error is OK all in all

File: tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+            """
+module Test
+    
+type M() =
+    static member A(n: System.Nullable<'T>) = ()
+    static member A(n: System.Nullable<int>) = ()
+    static member A(n: int) = ()
+
+let test() =
+    M.A(System.Nullable 3)
```
Comment (dsyme) [inline tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs:258]: Can we test execution, i.e. that the right overload got chosen in each case please?

File: src/Compiler/Checking/MethodCalls.fs (hunk 6 above)
Comment (NinoFloris) [reply src/Compiler/Checking/MethodCalls.fs:370]: Right, this keeps all the TDC work in one place, even though it's slightly inconsistently specced. It felt more natural to maintain this case alongside the other conversions.

File: tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+    M.A(System.Nullable 3.)
+    M.A(3)
+"""
+
+    [<Test>]
+    let ``Picking overload for typar does not favor any form of System.Nullable nor produce ambiguity warnings``() =
```
Comment (dsyme) [inline tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs:264]: Good test

File: tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+let test(x: 'T) =
+    M.A(x)
+"""
+         FSharpDiagnosticSeverity.Warning
+         64
+         (11, 5, 11, 11)
+         """This construct causes code to be less generic than indicated by the type annotations. The type variable 'T has been constrained to be type 'int'."""
+
+    [<Test>]
+    let ``Ambiguous overload for typar does not pick System.Nullable<'T>``() =
```
Comment (dsyme) [inline tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs:284]: Good test

File: tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs (hunk 9 above)
Comment (NinoFloris) [reply tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs:258]: Will add

File: tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+
+Candidates:
+ - static member M.A: n: System.Nullable<'T> -> unit when 'T: (new: unit -> 'T) and 'T: struct and 'T :> System.ValueType
+ - static member M.A: n: float -> unit
+ - static member M.A: n: int -> unit"""
+    
+    [<Test>]
+    let ``Passing an argument in nested method call property setter works``() =
+        CompilerAssert.CompileLibraryAndVerifyILWithOptions([|"--optimize-"|],
+            """
```
Comment (dsyme) [inline tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs:312]: It's important to have an actual execution test for this one as well (I may have missed it)

Comment (NinoFloris) [reply tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs:312]: Alright, it's indeed only testing the fixed IL, I'll add an execution test (not sure what it adds in this case though).

File: tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs [hunk 13]
```diff
@@ -0,0 +1,351 @@
...
+
+    [<Test>]
+    let ``Picking overload for typar does not favor any form of System.Nullable nor produce ambiguity warnings``() =
+        CompilerAssert.TypeCheckSingleError
+            """
+module Test
+    
+type M() =
+    static member A(n: System.Nullable<'T>) = ()
+//    static member A(n: System.Nullable<float>) = ()
```
Comment (NinoFloris) [inline tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs:271]: Ah, I did want to add, uncommenting this overload does force an ambiguity warning, here and on main, no defaulting to int. Just to confirm, that is what we want? (I also wouldn't quite know how to stop it)

File: src/Compiler/Checking/MethodCalls.fs (hunk 5 above)
Comment (dsyme) [reply src/Compiler/Checking/MethodCalls.fs:295]: OK, I understand, thank you

File: tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs (hunk 13 above)
Comment (dsyme) [reply tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs:271]: I'm a bit surprised by this
* With the extra overload, we get `A unique overload for method 'A' could not be determined`
* Without the extra overload we get a resolution to the `int` overload and `This construct causes code to be less generic than ....`

The same applies if the overload is `Nullable<char>`, e.g. the simpler:
```
type M() =
//    static member A(n: System.Nullable<char>) = printfn "1"
    static member A(n: System.Nullable<int>) = printfn "3"
    static member A(n: int) = printfn "4"
let test(x: 'T) =
    M.A(x)

test 1
```

I can see that extra warnings are being produced in each case and may affect resolution. However I don't understand why the extra overload causes this to happen - why ambiguity in the case of 3 and not 2? It feels like a flaw.

Comment (NinoFloris) [reply tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs:271]: Just to be sure there is no confusion, this is not a change that this PR introduced https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AbEAzAzmgFxAEsMAfAgTwAcYACAWQAoBKOgXgFgAoOvu3AQCGBYmDoBbGBOAwodAIJMAdiDoBlSoOkA6AHIBXDBiHAMMADxgAFkKgA+NuzrUoxZQWzK6AIgCMPjz8AsKi4lIycooqapraEvpGJmaW7gSOHC5uHl6+AMyBvPyCImKS0rLySqp0aU5Zabk+KIXmBHQEMIJMCGoA5AAqTkH8DDpKCCxAA

The reason this happens is that when we're in the later stage of overload resolution and we known nothing about the caller arg we fall through to https://github.com/dotnet/fsharp/pull/13673/files#diff-33f19e1cff6e1df09e36df423722449340f685c5c08f041543334c173cc259f8R373
At that point we don't mark these nullable candidates as using TDC and fail to trigger the resolution rule that would end up preferring int over any of the Nullable candidates.

Instead we rely entirely on the ad-hoc compareArg rule for Nullable https://github.com/dotnet/fsharp/blob/main/src/Compiler/Checking/ConstraintSolver.fs#L3208. This one works for the int and Nullable<int> (and even Nullable<'T>) selection case as we can dismiss all the other candidates. This fails to apply when underlying types like char and other non equivalent types are mixed in (with `enforceNullableOptionalsKnownTypes` forcing these to be in the list), leaving no best method.

After fixing that by having these candidates hit the TDC rule instead we get the same method printed twice as a candidate for the test `Ambiguous overload for typar does not pick System.Nullable<'T>` because we now have two warnings for this candidate. 

One being the diag stating it uses a builtin conversion and another stating the typar was not a match.

I can filter out DiagnosticWithText if there is more than one error here https://github.com/dotnet/fsharp/blob/main/src/Compiler/Checking/ConstraintSolver.fs#L3351 or maybe in the driver close to the output is better? (I don't quite understand why candidate method slots need to be duplicated per error).

I could also create a TDC case that produces no warning that will be used in this `enforceNullableOptionalsKnownTypes` branch.

Please advice :)

Comment (NinoFloris) [reply tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs]: For now I've done the latter https://github.com/dotnet/fsharp/pull/13673/commits/a84386ac0419dcb01a31851d13aa73b29254d122

I still think it's incomplete as I would want to get TDC warnings for the compiler picking a Nullable<'T> overload for 'T as well. But this needs these warning adjustments, I'm not so sure which way to go there.

File: src/Compiler/Checking/MethodCalls.fs (hunk 7 above)
Comment (NinoFloris) [reply src/Compiler/Checking/MethodCalls.fs:1524]: All this logic being so close to optional adjustments originates from FS-1075 being restricted to optional arguments at some point. Near completion it was expanded to cover all arguments but this code was never really moved to cover all arguments, which resulted in issues like #12994. 

Given `AdjustCallerArgExpr` runs for any argument after all fine-grained adjustments (which then unconditionally runs `AdjustExprForTypeDirectedConversions`) I chose to move it all to be colocated with the other TDC logic, just like I've done in that commit for the type checks in `AdjustRequiredTypeForTypeDirectedConversions`. 

The relevant commit is https://github.com/dotnet/fsharp/pull/13673/commits/da5d49b9a593d824782502f416c4b66c02e20bb5 and contains all changes pertaining to this.

File: src/Compiler/Checking/MethodCalls.fs (hunk 8 above)
Comment (NinoFloris) [reply src/Compiler/Checking/MethodCalls.fs:1495]: See https://github.com/dotnet/fsharp/pull/13673#discussion_r949553104

File: tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs (hunk 13 above)
Comment (dsyme) [reply tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs]: OK, I understand now.  I think don't touch this in this PR - so undo https://github.com/dotnet/fsharp/commit/a84386ac0419dcb01a31851d13aa73b29254d122 - let's get it green and have the primary fixes in the pipe

Comment (NinoFloris) [reply tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs:274]: Alright, removed and opened #13731

File: tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs (hunk 12 above)
Comment (NinoFloris) [reply tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs:315]: Added one

File: tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs (hunk 9 above)
Comment (NinoFloris) [reply tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs]: Added one

File: tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs [hunk 14]
```diff
@@ -0,0 +1,373 @@
...
+let test() =
+    M.A(System.Nullable 3.) = 1 |> assertTrue
+    M.A(System.Nullable 3) = 2 |> assertTrue
+    M.A(3) = 3 |> assertTrue
+    
+test()
+        """ []
+
+    [<Test>]
+    let ``Picking overload for typar does not favor any form of System.Nullable nor produce ambiguity warnings``() =
```
Comment (dsyme) [inline tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs:267]: Is the name of this test still correct?

Comment (dsyme) [reply tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs:267]: Also could you add the cases

* With the `Nullable<float>` overload
* With the `Nullable<float>` overload and without `Nullable<'T>` overload

File: tests/fsharp/Compiler/Language/OptionalInteropTests.fs [hunk 15]
```diff
@@ -6,15 +6,14 @@ open System.Collections.Immutable
...
-open FSharp.Compiler.Diagnostics
 open Microsoft.CodeAnalysis
 
 [<TestFixture>]
 module OptionalInteropTests =
 
-    [<Test>]
-    let ``C# method with an optional parameter and called with an option type should compile`` () =
+    [<TestCase("5.0")>]
+    [<TestCase("latest")>]
```
Comment (dsyme) [inline tests/fsharp/Compiler/Language/OptionalInteropTests.fs:15]: Should `preview` be added here?

File: tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs (hunk 14 above)
Comment (NinoFloris) [reply tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs:267]: As far as it's compatible with the current definition of 'producing ambiguity warnings' I'd say yes. I also opened an issue that tracks the rest.

File: tests/fsharp/Compiler/Language/OptionalInteropTests.fs (hunk 15 above)
Comment (NinoFloris) [reply tests/fsharp/Compiler/Language/OptionalInteropTests.fs:15]: I thought about it, that'd be fine. The commit that introduces this change specifically mentions it's to test the compiler pre and post FS-1093 additional conversions but preview would be compatible with that wording.

File: src/Compiler/Checking/MethodCalls.fsi [hunk 16]
```diff
@@ -119,7 +119,7 @@ type CallerArgs<'T> =
 /// has been used in F# code
 [<RequireQualifiedAccess>]
 type TypeDirectedConversionUsed =
-    | Yes of (DisplayEnv -> exn)
+    | Yes of (DisplayEnv -> exn) * isDoubleConversion: bool
```
Comment (dsyme) [inline src/Compiler/Checking/MethodCalls.fsi:122]: The name `DoubleConversion` is problematic - does it mean "int to double" or "there are two conversions".

Also could this better be an integer indicating the number of conversion rules applied?

File: src/Compiler/Checking/CheckExpressions.fs [hunk 17]
```diff
@@ -5385,7 +5385,7 @@ and TcAdjustExprForTypeDirectedConversions (cenv: cenv) (overallTy: OverallTy) a
     let g = cenv.g
 
     match overallTy with
-    | MustConvertTo (_, reqdTy) when g.langVersion.SupportsFeature LanguageFeature.AdditionalTypeDirectedConversions ->
+    | MustConvertTo (isMethodArg, reqdTy) when g.langVersion.SupportsFeature LanguageFeature.AdditionalTypeDirectedConversions || (g.langVersion.SupportsFeature LanguageFeature.NullableOptionalInterop && isMethodArg) ->
```
Comment (dsyme) [inline src/Compiler/Checking/CheckExpressions.fs:5388]: Thinking aloud: The conditional here is now quite complex - maybe it would have been better originally to just always call `AdjustExprForTypeDirectedConversions` and have the conditions on the specific rules in that method.  

No need to change anything here though.

**Comment by dsyme** [comment]
This is looking good - a couple of minor nits

File: src/Compiler/Checking/MethodCalls.fsi (hunk 16 above)
Comment (NinoFloris) [reply src/Compiler/Checking/MethodCalls.fsi:122]: Fair point, something like `isTwoStepConversion` would be a worse sounding but unambiguous name.

Regardless of naming this is a kludge to fix the immediate issue. I would suggest that the PR that fixes #13731 and fully resolves #13437 would introduce richer types like you did in #13063. that would be where we could capture potentially as a linked list (or just an integer indeed) all the conversions done.

Comment (dsyme) [reply src/Compiler/Checking/MethodCalls.fsi:122]: Yes `isTwoStepConversion` please

File: tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs (hunk 14 above)
Comment (NinoFloris) [reply tests/fsharp/Compiler/Language/TypeDirectedConversionTests.fs:267]: Added a failing test for https://github.com/dotnet/fsharp/issues/13731

File: tests/fsharp/Compiler/Language/OptionalInteropTests.fs (hunk 15 above)
Comment (NinoFloris) [reply tests/fsharp/Compiler/Language/OptionalInteropTests.fs]: Changed to preview here too

File: src/Compiler/Checking/MethodCalls.fsi (hunk 16 above)
Comment (NinoFloris) [reply src/Compiler/Checking/MethodCalls.fsi]: Renamed

**Comment by NinoFloris** [comment]
Ok so as mentioned offline, the only thing that's left to decide on is whether we should actually begin to emit the 'builtin conversion' warning under test here for Nullable<'T> https://github.com/dotnet/fsharp/pull/13673/commits/88ae3f8fafe2840db1d9125af4ababe199901cd8

The strongest argument for this change is that it brings those conversions in line with the other supported conversions.

The downside for this implementation in particular is that it will only emit a warning when full type info is available, basically until this problem https://github.com/dotnet/fsharp/pull/13673#discussion_r949358350 is tackled (captured in https://github.com/dotnet/fsharp/issues/13731) we'll have cases like 'T being converted to Nullable<'T> where the compiler won't warn yet. I think that's acceptable but maybe we want to do it all in one go.

@dsyme please let me know, I'd like to get this off my list :)

**Comment by dsyme** [comment]
> The downside for this implementation in particular is ...

I think we won't do this as part of this PR, partly because emitting the warning (even optional) would take it from being a fix to a possible RFC.

So I'll review now, and accept as is.  Thank you sooooo much for your detailed work here.

**Review by dsyme — APPROVED** [review]

**Review by vzarytovskii — APPROVED** [review]

**Comment by vzarytovskii** [comment]
@NinoFloris thanks a lot for this!

**Comment by dsyme** [comment]
@NinoFloris pointed out provately that this does emit the FS3389 off-by-default warning in an additional case, seen here https://github.com/dotnet/fsharp/pull/13673/files#diff-0020979b647768e28019703c22e81771d97b9e9b26eb7c3cb518c692e6596beaR264
[2:25](https://fsharp.slack.com/archives/DJC7FM64X/p1661174715549369)

This is ok and I consider this a bug fix as it's definitely a type-directed conversion outside the scope of optional-arg interop (the argument is not labelled optional in metadata, it just has explicit type Nullable).   

Additionally, the presence of the warning doesn't affect overload resolution because [this rule](https://github.com/dotnet/fsharp/pull/13673/files#diff-87bb55bb28da0160923c21cb7f0146311651517335a302a8a0d36692a797d639R3208) would already take this into account in overload resolution 

In short, all OK, I'm just noting this in case it comes up later
